File: python/src/dashgps/writers/gpxw.py
"""GPX 1.1 writer. spec/21-outputs.md.

One <trkseg> per non-glitch run, so a consumer that draws segments gets the recording gaps right
without having to know about dt_s.
"""

from ..fmt import P_ALT, P_ANGLE, P_LATLON, P_SPEED, fixed, iso_z, xml_text

NS = "http://www.topografix.com/GPX/1/1"
EXT_NS = "https://github.com/dashgps/dashgps/ns/1"


def write(out, group, ctx):
    version = ctx.get("version", "0.0.0")
    naive = ctx.get("time_is_naive", True)
    out.append('<?xml version="1.0" encoding="UTF-8"?>\n')
    out.append(
        '<gpx version="1.1" creator="dashgps ' + xml_text(version) + '" xmlns="' + NS
        + '" xmlns:dashgps="' + EXT_NS + '">\n'
    )
    if naive:
        out.append(
            "  <!-- Times are the camera's own clock, which carries no timezone. They are\n"
            "       written with a Z suffix because GPX requires one, but they are NOT UTC\n"
            "       unless dashgps was run with --tz-offset. -->\n"
        )
    out.append("  <metadata>\n")
    out.append("    <name>" + xml_text(group.label) + "</name>\n")
    if group.points:
        out.append("    <time>" + iso_z(group.points[0].t) + "</time>\n")
    out.append("  </metadata>\n")

    segs = [r for r in group.runs if not r.glitch]
    if segs:
        _trk(out, group, segs, group.label)
    if ctx.get("include_glitch"):
        bad = [r for r in group.runs if r.glitch]
        if bad:
            _trk(out, group, bad, group.label + " (glitch)")
    out.append("</gpx>\n")


def _trk(out, group, runs, name):
    out.append("  <trk>\n")
    out.append("    <name>" + xml_text(name) + "</name>\n")
    for r in runs:
        out.append("    <trkseg>\n")
        for i in range(r.start, r.end):
            p = group.points[i]
            out.append(
                '      <trkpt lat="' + fixed(p.lat, P_LATLON)
                + '" lon="' + fixed(p.lon, P_LATLON) + '">\n'
            )
            a = fixed(p.alt_m, P_ALT)
            if a:
                out.append("        <ele>" + a + "</ele>\n")
            out.append("        <time>" + iso_z(p.t) + "</time>\n")
            s = fixed(p.speed_kmh, P_SPEED)
            h = fixed(p.heading_deg, P_ANGLE)
            if s or h:
                out.append("        <extensions>\n")
                if s:
                    out.append("          <dashgps:speed_kmh>" + s + "</dashgps:speed_kmh>\n")
                if h:
                    out.append("          <dashgps:heading_deg>" + h + "</dashgps:heading_deg>\n")
                out.append("        </extensions>\n")
            out.append("      </trkpt>\n")
        out.append("    </trkseg>\n")
    out.append("  </trk>\n")

File: js/src/writers/gpxw.js
// GPX 1.1 writer. spec/21-outputs.md  (mirrors python/src/dashgps/writers/gpxw.py)

import { P_ALT, P_ANGLE, P_LATLON, P_SPEED, fixed, isoZ, xmlText } from "../fmt.js";

const NS = "http://www.topografix.com/GPX/1/1";
// A URN, not a URL: an XML namespace is an identifier and is never fetched, and this one must
// stay stable if the project ever moves. spec/21-outputs.md
const EXT_NS = "urn:dashgps:gpx:1";

export function write(out, group, ctx) {
  const version = ctx.version || "0.0.0";
  const naive = ctx.timeIsNaive !== false;
  out.push('<?xml version="1.0" encoding="UTF-8"?>\n');
  out.push('<gpx version="1.1" creator="dashgps ' + xmlText(version) + '" xmlns="' + NS +
           '" xmlns:dashgps="' + EXT_NS + '">\n');
  if (naive) {
    out.push("  <!-- Times are the camera's own clock, which carries no timezone. They are\n" +
             "       written with a Z suffix because GPX requires one, but they are NOT UTC\n" +
             "       unless dashgps was run with --tz-offset. -->\n");
  }
  out.push("  <metadata>\n");
  out.push("    <name>" + xmlText(group.label) + "</name>\n");
  if (group.points.length) out.push("    <time>" + isoZ(group.points[0].t) + "</time>\n");
  out.push("  </metadata>\n");

  const segs = group.runs.filter((r) => !r.glitch);
  if (segs.length) trk(out, group, segs, group.label);
  if (ctx.includeGlitch) {
    const bad = group.runs.filter((r) => r.glitch);
    if (bad.length) trk(out, group, bad, group.label + " (glitch)");
  }
  out.push("</gpx>\n");
}

function trk(out, group, runs, name) {
  out.push("  <trk>\n");
  out.push("    <name>" + xmlText(name) + "</name>\n");
  for (const r of runs) {
    out.push("    <trkseg>\n");
    for (let i = r.start; i < r.end; i++) {
      const p = group.points[i];
      out.push('      <trkpt lat="' + fixed(p.lat, P_LATLON) + '" lon="' +
               fixed(p.lon, P_LATLON) + '">\n');
      const a = fixed(p.altM, P_ALT);
      if (a) out.push("        <ele>" + a + "</ele>\n");
      out.push("        <time>" + isoZ(p.t) + "</time>\n");
      const s = fixed(p.speedKmh, P_SPEED);
      const h = fixed(p.headingDeg, P_ANGLE);
      if (s || h) {
        out.push("        <extensions>\n");
        if (s) out.push("          <dashgps:speed_kmh>" + s + "</dashgps:speed_kmh>\n");
        if (h) out.push("          <dashgps:heading_deg>" + h + "</dashgps:heading_deg>\n");
        out.push("        </extensions>\n");
      }
      out.push("      </trkpt>\n");
    }
    out.push("    </trkseg>\n");
  }
  out.push("  </trk>\n");
}
